net/ws.ts

```diff
@@ -0,0 +1,350 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+import { Buffer, Writer, Conn } from "deno";
+import { ServerRequest } from "./http.ts";
+import { BufReader, BufWriter } from "./bufio.ts";
+import { readLong, readShort, sliceLongToBytes } from "./ioutil.ts";
+import { Sha1 } from "./sha1.ts";
+
+export const OpCodeContinue = 0x0;
+export const OpCodeTextFrame = 0x1;
+export const OpCodeBinaryFrame = 0x2;
+export const OpCodeClose = 0x8;
+export const OpcodePing = 0x9;
+export const OpcodePong = 0xa;
+
+export type WebSocketEvent =
+  | string
+  | Uint8Array
+  | WebSocketCloseEvent
+  | WebSocketPingEvent
+  | WebSocketPongEvent;
+
+export type WebSocketCloseEvent = {
+  code: number;
+  reason?: string;
+};
+
+export function isWebSocketCloseEvent(a): a is WebSocketCloseEvent {
+  return a && typeof a["code"] === "number";
+}
+
+export type WebSocketPingEvent = ["ping", Uint8Array];
+
+export function isWebSocketPingEvent(a): a is WebSocketPingEvent {
+  return Array.isArray(a) && a[0] === "ping" && a[1] instanceof Uint8Array;
+}
+
+export type WebSocketPongEvent = ["pong", Uint8Array];
+
+export function isWebSocketPongEvent(a): a is WebSocketPongEvent {
+  return Array.isArray(a) && a[0] === "pong" && a[1] instanceof Uint8Array;
+}
+
+export class SocketClosedError extends Error {}
+
+export type WebSocketFrame = {
+  isLastFrame: boolean;
+  opcode: number;
+  mask?: Uint8Array;
+  payload: Uint8Array;
+};
+
+export type WebSocket = {
```

Comment:
Is it better to declare this explicitly as an Interface?

And a general question about this repo, since we are talking about the standard deno repo, would it make sense to group all the interfaces under one folder to let different implementations import the interface only as an API? I believe talking about this would make more sense once deno is stable enough to have a stable API.

Reply:
> My first thought is that it seems like it would be a bit confusing to newcomers
true
It might be too much to include one /contracts folder under the root of this repo. Yet it would be nice to have an explicit access to a contract if someone decided to start his own implementation. I would say an interface or an abstract class  per module's folder in this repo would be a minimal solid contract